[ESP32 MAX] Configuration for ESP32 with 16MB Flash

tonhuisman opened the pull request:
Added configurations for ESP32 with 16 MB Flash (and also an _ETH like other _ETH configurations)
As ESP32 with 16MB Flash are quite affordable now, I bought [this ESP32 with 16MB](https://www.tinytronics.nl/shop/nl/communicatie/wi-fi/wemos-lolin-d32-pro-v2-esp32-ch340c-16mb-flash-8mb-psram) just to see if I could cram all plugins into a single build.
I couldn't test all plugins, obviously, but it compiles and boots on the hardware. 😃 

Chose for a setup of 4 MB Sketch size, and 2MB or nearly 8MB of file storage (LittleFS filesystem) or 1MB of file storage (SPIFFS), and ditto for ETH builds.
- [x] OTA update allows a 4MB .bin file
- [x] Max plugin count for now 'limited' to 125
- [x] All plugins and controllers usable on ESP32 included

tonhuisman committed:
[ESP32 MAX] Configuration for ESP32 with 16MB Flash

tonhuisman committed:
[ESP32 MAX] Include in travis build

TD-er commented:
8M of SPIFFS storage, isn't that terribly slow?

tonhuisman commented:
> 
> 
> 8M of SPIFFS storage, isn't that terribly slow?

I was thinking of enabling LittleFS, but that currently isn't supported for ESP32, so I hope to add that in a next commit. But I have to investigate that first. And I'm not sure if the current ESP32 SPIFFS implementation isn't actually LittleFS in disguise. (as it's using a / as prefix for files, and LittleFS does support directories...)

TD-er commented:
ESP32 still uses SPIFFS, but on ESP32 it does support "folders".
As in it does support file names with a `/` in them and iterate over them as if it is a folder.

Little FS is extremely fast when reading, but writing -as we do it- is extremely slow.
Little FS cannot update a part of a file. It has to rewrite all from the point where you change things.
Thus saving settings on ESP8266 may take 2 - 3 seconds on each save. On ESP32 it will be even slower as the config.dat file is twice as large.
So "support for Little FS" is a bit more than just including support for the file system.

I tested SPIFFS on larger flash chips like the Wemos D1 mini pro (the one with 16M flash) and as soon as I got over roughly 2M of SPIFFS size, it all became very slow on every file access.
So loading the web pages was extremely slow on 14M SPIFFS (unworkable)
I did include some optimizations since but still, SPIFFS is rather slow when it becomes larger.
I guess you could experiment a bit with it, but I don't think you should go over 2M for SPIFFS.
Maybe the ESP32 is just that bit faster to maybe allow for 4M flash, but I strongly suggest to test it by for example loading web pages like the controller settings page (thus reading lots from the settings file)

tonhuisman committed:
[ESP32 MAX] Extend configurations with 2MB and 4MB SPIFFS, 4MB default

tonhuisman committed:
[ESP32 MAX] Enable SD-card support (and fix compilation errors)

tonhuisman commented:
I tested the different SPIFFS setups, 2MB, 4MB and (nearly) 8MB, and although 8MB is somewhat slower in displaying the web pages, it's still very usable/workable. I did set up the 2MB SPIFFS as the base environment, and replaced the 8MB with the 4MB setup for Travis to build, suggesting that as the general use bin, and also used that as the basis for the ETH configuration.
Haven't tried LittleFS yet.
The partition layouts are these:


- [x] **max_ESP32_16M2M:**


![ESP32-16M2M_Partitions](https://user-images.githubusercontent.com/1922393/110212380-65548280-7e9b-11eb-905e-a0ab5ff7c13e.png)


- [x] **max_ESP32_16M8M:**


![ESP32-16M8124k_Partitions](https://user-images.githubusercontent.com/1922393/110212414-8321e780-7e9b-11eb-9a04-26b3b3d0dac6.png)

I _deliberately_ positioned the SPIFFS at the end of the flash, so a) you will need initially to flash over USB (using the factory bin) to get the correct partition setup, and b) when switching a layout, the other configurations will (probably) stay intact. Switching SPIFFS size is probably only useful during testing/experimenting.

**Attention:** When 'switching' SPIFFS size it is **required** to first flash the factory bin, or else it will use the previous partition layout, possibly causing issues when you have to re-flash for some reason. Updates can be done via OTA.

~~For those interested, here are my local builds (100 plugins are available) for the 3 configurations without hardware ethernet support: (includes both 'generic' and '-factory' bin files)~~
(Removed as the SPIFFS file-system proved to not work properly on ESP32 with > 1 MB assigned, now switched to LittleFS and only 2MB and 8MB builds)

Partial build output:
```
RAM:   [==        ]  17.8% (used 58408 bytes from 327680 bytes)
Flash: [=====     ]  54.6% (used 2289382 bytes from 4194304 bytes)
```
So there is some room for more plugins 😃

tonhuisman commented:
>  but writing -as we do it- is extremely slow.

@TD-er what is the reason configuration is written one byte at a time instead of just writing the entire chunk of data, or in blocks of 256 bytes (to be able to satisfy the watchdog)?
ATM, I'm seeing write errors on any SPIFFS size > 1 MB (not yet included in the environments), and that's rather unexpected and of course undesirable.

tonhuisman committed:
[ESP32 MAX] Add configuration with 1MB SPIFFS

TD-er commented:
Writing to the file is done per byte, but writing to the flash itself is buffered so it is not written that often :)
I am not entirely sure anymore why it is written to the stream a byte at a time, but I guess there was a good reason for it.
I think (not sure) that it may have been unclear when the data is actually flushed to flash by SPIFFS and this allows to keep the watchdog asleep too on slow flash chips.

Please note that changing to LittleFS may require us to split the config.dat file into several smaller files and create an import/export wrapper to be able to handle it as a single file.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/ESP32_16MB-flash-max-plugins

TD-er commented on patches/check_puya_patch:
```diff
@@ -2,7 +2,7 @@
 PATCH=puya_core2_4_x.patch
 
-for DIR in ~/.platformio/packages/framework-arduinoespressif8266*/; do
+for DIR in ~/.platformio/packages/framework-arduinoespressif8266/ ~/.platformio/packages/framework-arduinoespressif8266*/; do
```
Why this change?
Shouldn't `*` also match the empty chars?

tonhuisman replied:
I didn't touch that file or line... at least I can't find it in the Files changed listing above, and I have no intention of changing it 🤔

TD-er replied:
Maybe it was already part of the just merged Actions build.
I will check to see if it really makes a difference.

tonhuisman replied:
It is part of #3525

tonhuisman committed:
[ESP32 MAX] Mixed up partitions fixed

TD-er commented on platformio_esp32_envs.ini:
```diff
@@ -133,3 +133,35 @@ build_flags               = ${env:test_ESP32_4M316k.build_flags} -DHAS_ETHERNET
+board_build.partitions    = esp32_partition_app4096k_spiffs4096k.csv
+
+; A Lolin D32 PRO with 16MB Flash, allowing 4MB sketch size, and the rest (nearly 8MB) assigned to file storage
+[env:max_ESP32_16M8124k]
```
Maybe rename this one to `env:max_ESP32_16M8M` ?
Don't forget in the Python script for testing resulting file size.

Or did you use the 'k' notation as it isn't a full 8192k size?

tonhuisman replied:
> Or did you use the 'k' notation as it isn't a full 8192k size?

Yes, that's the reason.

OT:
I'm now testing with LittleFS (using lorol/LITTLEFS.git as the redirected LittleFS_esp32 throws a weird error during compilation) and I do see it's a tad slow during save, but as SPIFFS is deprecated, I'm thinking of only offering LittleFS variants for the 16MB ESP32's, as there's no backward compatibility required here.

tonhuisman replied:
> Maybe rename this one to `env:max_ESP32_16M8M` ?

I've now renamed it, and switched to LittleFS for all 16MB environments, will be in the next commit. Much easier to pronounce, and because of the overhead none of the file systems are offering the mentioned size anyway.

tonhuisman committed:
[ESP32 MAX] Switch to LittleFS, only 16M2M and 16M8M configs + ETH

tonhuisman commented:
For those interested in trying these, click on a Details link in the build list, find the Artifacts button on the right, and download the Binaries archive, that contains an archive per build environment. (Very convenient 👍)

tonhuisman committed:
[ESP32 MAX] Fix build issue, case-sensitive filename

TD-er commented:
Isn't LittleFS too slow in saving settings?

tonhuisman commented:
> Isn't LittleFS too slow in saving settings?

Not really, but is isn't fast either.

TD-er commented:
Like I said, for LittleFS we really should split the Config.dat file.
Maybe a good idea for a new PR (when this is merged) to consider the Config.dat file as source and split it to several separate files (in a folder which will be hidden from the file system browser)
When the Config.dat file is then requested, we can either stitch it together from the separate files, or first need to recreate it to access it via the file browser (or a fetch URL)

tonhuisman commented:
Some logs:
Saving a controller:
```
6374886 : Info   : SaveToFile: free stack: 7144
6376021 : Info   : FILE : Saved /config.dat
6376022 : Info   : SaveToFile: free stack after: 7144
6376030 : Info   : SaveToFile: free stack: 6648
6377111 : Info   : FILE : Saved /config.dat
6377112 : Info   : SaveToFile: free stack after: 6648
6377113 : Info   : SaveToFile: free stack: 7112
6378313 : Info   : FILE : Saved /config.dat
6378314 : Info   : SaveToFile: free stack after: 7112
6378322 : Info   : SaveToFile: free stack: 6744
6378375 : Info   : FILE : Saved /security.dat
6378376 : Info   : SaveToFile: free stack after: 6744
```
Saving a device:
```
6513169 : Info   : SaveToFile: free stack: 6568
6514069 : Info   : FILE : Saved /config.dat
6514070 : Info   : SaveToFile: free stack after: 6568
6514075 : Info   : SaveToFile: free stack: 5832
6514978 : Info   : FILE : Saved /config.dat
6514979 : Info   : SaveToFile: free stack after: 5832
6514982 : Info   : SaveToFile: free stack: 7128
6515878 : Info   : FILE : Saved /config.dat
6515879 : Info   : SaveToFile: free stack after: 7128
6515880 : Info   : SaveToFile: free stack: 7112
6517081 : Info   : FILE : Saved /config.dat
6517082 : Info   : SaveToFile: free stack after: 7112
6517086 : Info   : SaveToFile: free stack: 6744
6517130 : Info   : FILE : Saved /security.dat
6517131 : Info   : SaveToFile: free stack after: 6744
```

NB: It's a bit confusing that config.dat seems to be saved ~~twice~~ multiple times.

TD-er commented:
Yep that does look a bit overkill.
Maybe also add the offset to where it is written?

tonhuisman commented:
Updated log with offset and size during save of a task:
~~~
31488 : Info   : SaveToFile: free stack: 6568
32879 : Info   : FILE : Saved /config.dat offset: 32768 size: 472
32880 : Info   : SaveToFile: free stack after: 6568
32889 : Info   : SaveToFile: free stack: 5832
34263 : Info   : FILE : Saved /config.dat offset: 33792 size: 816
34264 : Info   : SaveToFile: free stack after: 5832
34270 : Info   : SaveToFile: free stack: 7128
35629 : Info   : FILE : Saved /config.dat offset: 32768 size: 472
35631 : Info   : SaveToFile: free stack after: 7128
35632 : Info   : SaveToFile: free stack: 7112
37467 : Info   : FILE : Saved /config.dat offset: 0 size: 3000
37468 : Info   : SaveToFile: free stack after: 7112
37475 : Info   : SaveToFile: free stack: 6744
37546 : Info   : FILE : Saved /security.dat offset: 1024 size: 128
37547 : Info   : SaveToFile: free stack after: 6744
~~~

tonhuisman committed:
[EPS32 MAX] Extend logging during configuration save

tonhuisman commented:
It's not just the saving to file part, fetching the data from the page after clicking the Submit button also seems quite slow:
~~~
848644 : Info   :  Webserver args: 0: 'index' length: 1
853085 : Info   :  Webserver args: 0: 'index' length: 1 1: 'protocol' length: 1 2: 'locatecontroller' length: 1 3: 'controllerip' length: 12  ...
853090 : Info   : SaveToFile: free stack: 7080
~~~
(trimmed out some of the content for readability)

TD-er commented:
OK, so it does look like it is indeed saving some task specific data and then the SettingsStruct at last
So that's indeed what you would expect and also what is especially what LittleFS can't handle very well.
Just run the command "meminfodetail" on the ESP (does also work on ESP8266) and then load the sysinfo page and scroll to the bottom to see how the files are structured and with what offsets.

TD-er commented:
> It's not just the saving to file part, fetching the data from the page after clicking the Submit button also seems quite slow:
> 
> ```
> 848644 : Info   :  Webserver args: 0: 'index' length: 1
> 853085 : Info   :  Webserver args: 0: 'index' length: 1 1: 'protocol' length: 1 2: 'locatecontroller' length: 1 3: 'controllerip' length: 12  ...
> 853090 : Info   : SaveToFile: free stack: 7080
> ```
> 
> (trimmed out some of the content for readability)

Fetching after submit is waiting for the write to finish (which may take seconds).
Just reading from LittleFS is incredibly fast compared to SPIFFS.

tonhuisman commented:
> Just run the command "meminfodetail" on the ESP (does also work on ESP8266) and then load the sysinfo page and scroll to the bottom to see how the files are structured and with what offsets.

![Screenshot - 07_03_2021 , 22_43_42](https://user-images.githubusercontent.com/1922393/110255981-b4331280-7f96-11eb-956f-5bd505f9e3d3.png)

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

tonhuisman committed:
[ESP32_MAX] Add plugin P108 to ESP32 Max set

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

TD-er commented:
So the 8MB SPIFFS is not an issue when saving settings?

tonhuisman commented on platformio_esp32_envs.ini:
```diff
@@ -133,3 +133,37 @@ build_flags               = ${env:test_ESP32_4M316k.build_flags} -DHAS_ETHERNET
+board_upload.maximum_size = 4194304
+build_flags               = ${esp32_common.build_flags}  
+                            -DFEATURE_ARDUINO_OTA
+                            -DUSE_LITTLEFS
```
I've enabled LittleFS by default for this 16M Flash configuration, as SPIFFS was unreliable when configured > 1MB
8MB storage should probably only be used when really needed, default should be the 2MB configuration.

TD-er replied:
Hmm LittleFS still needs some patching in how the settings file is written.
Like I said before on LittleFS you can't update some portion somewhere in a file. All behind it has to be written again.
So saving may take quite a while (at least on ESP8266 when I tested it)

TD-er replied:
Also the text in the drawing should be changed to LittleFS, when it is being used, as that tipped me off into making my comment :)

tonhuisman replied:
> Like I said, for LittleFS we really should split the Config.dat file.
> Maybe a good idea for a new PR (when this is merged) to consider the Config.dat file as source and split it to several separate files (in a folder which will be hidden from the file system browser)
> When the Config.dat file is then requested, we can either stitch it together from the separate files, or first need to recreate it to access it via the file browser (or a fetch URL)

This is what you said before 👍

TD-er commented on tools/pio/copy_files.py:
```diff
@@ -23,6 +23,9 @@ def get_max_bin_size(env_name, file_suffix):
         if "factory" in file_suffix:
             # Factory bin files include a part which is not overwritten via OTA
             max_bin_size = max_bin_size + 65536
+    if "_16M8M" in env_name or "_16M2M" in env_name:
```
Maybe also check for "ESP32" here?

tonhuisman replied:
NP, can change later today, this felt more generic 😄

TD-er replied:
Yep, but you can't have 4M sketch for ESP8266, but you can have 16M of flash on ESP8266.

TD-er commented on src/src/Helpers/FS_Helper.h:
```diff
@@ -28,7 +28,7 @@
 #if defined(ESP32)
   #ifdef USE_LITTLEFS
-    #include "LittleFS.h"
+    #include "LITTLEFS.h"
```
Huh? Did they change the filename?
That also explains why it suddenly would not build anymore on builds without "beta" in the env. name.
Please also check on other PIO envs with "LittleFS in the env name to see if the capitalization has changed somewhere...

tonhuisman replied:
I've had to use another LittleFS implementation, as the 'original' one wouldn't play nice. This one is only used for these ESP32 builds.

TD-er replied:
Ah OK.
I just looked at the readme of that project just now and apparently it is in the process of being integrated into the main ESP32 Arduino core.

tonhuisman commented:
@TD-er Would it be useful to add a 16M512k SPIFFS configuration? (I've always found the 4M316k a bit odd value, but I understand why it is that size). Or is SPIFFS nearing the 'deprecated' status?

TD-er commented:
The 4M1M is working very well with SPIFFS, so I guess making it 16M1M makes more sense for SPIFFS.

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

tonhuisman committed:
[ESP32 MAX] Add 16M1M config (SPIFFS), rename 2M/8M (LittleFS) configs

tonhuisman committed:
[ESP32 MAX] Add ESP32_16M1M_ETH configuration, disable PS-RAM support

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

tonhuisman committed:
[ESP32 MAX] Enable most plugins up to P125 (even when not in PR yet)

tonhuisman committed:
[ESP32 MAX] Enable P109 now that it is in PR

TD-er commented on src/src/CustomBuild/define_plugin_sets.h:
```diff
@@ -1178,18 +1178,78 @@ To create/register a plugin, you have to :
     #define USES_P108   // DDS238-x ZN MODBUS energy meter (was P224 in the Playground)
   #endif
+  // #ifndef USES_P109
+  //   #define USES_P109   // ThermOLED (unconfirmed yet)
```
It is confirmed

tonhuisman replied:
I'll remove the comment later today

tonhuisman committed:
[ESP32 MAX] Clean up an obsolete comment

tonhuisman committed:
[ESP32 MAX] Add ENERGY and DISPLAY plugin sets, remove EXPERIMENTAL set

tonhuisman committed:
Merge branch 'mega' into feature/ESP32_16MB-flash-max-plugins

TD-er commented:
Ready for merge?

tonhuisman commented:
Yes, AFAICS I've done all I could, except for the split config files, but that's to be a new PR (and quite some work I estimate).
